File: frontend/src/types/task.ts
import type { ComplianceResponse } from "./compliance";
import type { TranscriptResponse } from "./transcript";
import type { EvaluationResponse, EvaluationResult } from "./evaluation";

export type TaskStatus =
  | "pending"
  | "queued_asr"
  | "processing_asr"
  | "extracting_frames"
  | "scanning_visual"
  | "correcting"
  | "evaluating"
  | "auditing"
  | "completed"
  | "failed";

export interface TaskSubmitResponse {
  task_id: string;
  status: TaskStatus;
  existing?: boolean;
}

export interface TaskProgress {
  current_chunk: number;
  total_chunks: number;
  percent: number;
}

export interface TaskStatusResponse {
  task_id: string;
  status: TaskStatus;
  progress: TaskProgress;
  result: EvaluationResponse | TranscriptResponse | ComplianceResponse | null;
  error: string | null;
}

export interface TaskResultsResponse {
  task_id: string;
  transcript: TranscriptResponse | null;
  evaluation: EvaluationResult | null;
  compliance: ComplianceResponse | null;
  has_audio: boolean;
  has_video: boolean;
  has_synthesis: boolean;
  keyframe_count: number;
  ocr_text_count: number;
  visual_event_count: number;
}

export interface TaskSummary {
  task_id: string;
  /** 用户重命名 > 纪要标题 > 原始文件名 */
  name: string;
  filename: string;
  created_at: string;
  status: TaskStatus;
  error: string | null;
  has_video: boolean;
  has_evaluation: boolean;
  has_compliance: boolean;
}

export interface TaskListResponse {
  tasks: TaskSummary[];
  /** 磁盘上的任务总数，大于 tasks.length 表示被截断 */
  total: number;
}

export interface TranscriptTextEdit {
  index: number;
  text_corrected: string;
}

File: backend/src/copernicus/schemas/task.py
from enum import StrEnum

from pydantic import BaseModel, Field, field_validator

from copernicus.schemas.compliance import ComplianceResponse
from copernicus.schemas.evaluation import EvaluationResponse, EvaluationResult
from copernicus.schemas.transcription import (
    MAX_SPEAKER_NAME_LEN,
    MAX_TRANSCRIPT_TEXT_LEN,
    TranscriptResponse,
)


class TaskStatus(StrEnum):
    PENDING = "pending"
    PROCESSING_ASR = "processing_asr"
    EXTRACTING_FRAMES = "extracting_frames"
    SCANNING_VISUAL = "scanning_visual"
    CORRECTING = "correcting"
    EVALUATING = "evaluating"
    AUDITING = "auditing"
    COMPLETED = "completed"
    FAILED = "failed"


class TaskSubmitResponse(BaseModel):
    task_id: str
    status: TaskStatus
    existing: bool = False


class TaskProgress(BaseModel):
    current_chunk: int = 0
    total_chunks: int = 0
    percent: float = 0.0


class TaskStatusResponse(BaseModel):
    task_id: str
    status: TaskStatus
    progress: TaskProgress
    result: EvaluationResponse | TranscriptResponse | ComplianceResponse | None = None
    error: str | None = None


class TaskResultsResponse(BaseModel):
    """Persisted results for a task (used for restoring state on page refresh)."""

    task_id: str
    transcript: TranscriptResponse | None = None
    evaluation: EvaluationResult | None = None
    compliance: ComplianceResponse | None = None
    has_audio: bool = False
    has_video: bool = False
    has_synthesis: bool = False
    keyframe_count: int = 0
    ocr_text_count: int = 0
    visual_event_count: int = 0


class TaskSummary(BaseModel):
    """历史任务列表条目。"""

    task_id: str
    name: str  # 优先级：用户重命名 > 纪要标题 > 原始文件名
    filename: str
    created_at: str
    status: TaskStatus
    error: str | None = None
    has_video: bool = False
    has_evaluation: bool = False
    has_compliance: bool = False


class TaskListResponse(BaseModel):
    tasks: list[TaskSummary]
    total: int  # 磁盘上的任务总数（大于 len(tasks) 表示被 limit 截断）


class TaskRenameRequest(BaseModel):
    name: str = Field(min_length=1, max_length=100)


class TranscriptTextEdit(BaseModel):
    index: int = Field(ge=0, description="句段在转写列表中的下标")
    text_corrected: str = Field(max_length=MAX_TRANSCRIPT_TEXT_LEN)


class TranscriptEditRequest(BaseModel):
    edits: list[TranscriptTextEdit] = Field(min_length=1, max_length=2000)


class SpeakerRenameRequest(BaseModel):
    """{原说话人: 新名称}；多个原名映射到同一新名称即合并说话人。"""

    renames: dict[str, str] = Field(min_length=1, max_length=50)

    @field_validator("renames")
    @classmethod
    def _clean_names(cls, renames: dict[str, str]) -> dict[str, str]:
        cleaned = {old: new.strip() for old, new in renames.items()}
        if any(not n or len(n) > MAX_SPEAKER_NAME_LEN for n in cleaned.values()):
            raise ValueError(f"说话人名称不能为空且不超过 {MAX_SPEAKER_NAME_LEN} 字符")
        return cleaned


class TranscriptUpdateResponse(BaseModel):
    updated: int
